src/gateway/sharding/shard_manager.rs

```diff
@@ -141,16 +141,18 @@ impl ShardManager {
         shard_init: u16,
         shard_total: NonZeroU16,
     ) -> Result<(), GatewayError> {
-        self.initialize(shard_index, shard_init, shard_total).await;
+        self.initialize(shard_index, shard_init, shard_total);
         loop {
             if let Ok(Some(msg)) =
                 timeout(self.wait_time_between_shard_start, self.manager_rx.next()).await
             {
                 match msg {
-                    ShardManagerMessage::Boot(shard_id) => self.boot(shard_id, false).await,
+                    ShardManagerMessage::Boot(shard_id) => self.queue_for_start(shard_id),
                     ShardManagerMessage::Quit(err) => return Err(err),
                 }
             }
+            let batch = self.queue.pop_batch();
+            self.checked_start(batch).await;
```

Comment:
Currently this is fine, but seems fragile. I worry that this will end up being called too often in future, if more ShardManagerMessages are added.

Reply:
Could you elaborate? I could put a check for if the queue is empty, which would require a linear scan of the buckets but would save creating an empty `Vec` each time.

Reply:
Well, `checked_start` already takes care of sleeping between shard batches. If restarts come in quickly, then they'll be caught and spaced out accordingly.

Reply:
Ah, that's what I was missing. This is fine then.